usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -4601,6 +4612,245 @@ static CK_RV rsa_ec_generate_keypair(STDLL_TokData_t * tokdata,
+    }
+
+    rc = override_key_attributes(tokdata, ktype, CKO_PUBLIC_KEY,
+                                 new_pPublicKeyTemplate,
+                                 new_ulPublicKeyAttributeCount);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s Dilithium override public key attributes failed with "
+                    "rc=0x%lx\n", __func__, rc);
+        goto error;
+    }
+
+    rc = override_key_attributes(tokdata, ktype, CKO_PRIVATE_KEY,
+                                 new_pPrivateKeyTemplate,
+                                 new_ulPrivateKeyAttributeCount);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s Dilithium override private key attributes failed with "
+                    "rc=0x%lx\n", __func__, rc);
+        goto error;
+    }
+
+    /* debug */
+    for (i = 0; i < new_ulPrivateKeyAttributeCount; i++) {
+        TRACE_INFO("%s gen priv attr type=0x%lx valuelen=0x%lx attrcnt=0x%lx\n",
+                   __func__, new_pPrivateKeyTemplate[i].type,
+                   new_pPrivateKeyTemplate[i].ulValueLen,
+                   new_ulPrivateKeyAttributeCount);
+    }
+
+    ep11_get_pin_blob(ep11_session,
+                      (ep11_is_session_object
+                       (pPublicKeyTemplate, ulPublicKeyAttributeCount)
+                       || ep11_is_session_object(pPrivateKeyTemplate,
+                                                 ulPrivateKeyAttributeCount)),
+                      &ep11_pin_blob, &ep11_pin_blob_len);
+
+    RETRY_START
+        rc = dll_m_GenerateKeyPair(pMechanism, new_pPublicKeyTemplate,
+                                   new_ulPublicKeyAttributeCount,
+                                   new_pPrivateKeyTemplate,
+                                   new_ulPrivateKeyAttributeCount,
+                                   ep11_pin_blob, ep11_pin_blob_len,
+                                   privkey_blob, &privkey_blob_len, spki,
+                                   &spki_len, ep11_data->target);
+    RETRY_END(rc, tokdata, sess)
+    if (rc != CKR_OK) {
+        rc = ep11_error_to_pkcs11_error(rc, sess);
+        TRACE_ERROR("%s m_GenerateKeyPair rc=0x%lx spki_len=0x%zx "
+                    "privkey_blob_len=0x%zx mech='%s'\n",
+                    __func__, rc, spki_len, privkey_blob_len,
+                    ep11_get_ckm(pMechanism->mechanism));
+        goto error;
+    }
+    TRACE_INFO("%s m_GenerateKeyPair rc=0x%lx spki_len=0x%zx "
+               "privkey_blob_len=0x%zx mech='%s'\n",
+               __func__, rc, spki_len, privkey_blob_len,
+               ep11_get_ckm(pMechanism->mechanism));
+
+    if (spki_len > MAX_BLOBSIZE || privkey_blob_len > MAX_BLOBSIZE) {
+        TRACE_ERROR("%s blobsize error\n", __func__);
+        rc = CKR_KEY_INDIGESTIBLE;
+        goto error;
+    }
+
+    memset(publblob, 0, sizeof(publblob));
+    memset(privblob, 0, sizeof(privblob));
+
+    memcpy(publblob, spki, spki_len);
+    publblobsize = spki_len;
+
+    memcpy(privblob, privkey_blob, privkey_blob_len);
```

Comment:
imo privkey_blob can (and should) be erased afterwards.

Reply:
Whats the reason for copying privkey_blob to privblob at all? Can't you just use the key blob from privkey_blob ? 
Same for the public blob. Why copy it?
Other key pair generating functions (e.g. RSA, EC, DH, DSA) in EP11 do not copy it, so why does this function?

Reply:
Good point ... in fact this code is not necessary. I removed it, retested, and saw no difference. The same piece of code also appears in rsa_ec_generate_keypair(), where it is also superfluous. I also removed the code there and retested with same result. So I'm now pushing the code change for Dilithium, but let's keep in mind to change rsa_ec_generate_keypair() accordingly in a separate pull request.